h264: replace usage of old BitReader with new

saxon-milton opened the pull request:
It turned into a much bigger job than I could imagine. As you can see, it's mostly mechanical changes. 
Updates #43.

saxon-milton committed:
h264: removed old BitReader from read.go

saxon-milton committed:
h264: started replacing usage of old BitReader

saxon-milton committed:
h264/read.go: made required changes to reading utilities and added some functions

Made some required changes to start removing old BitReader from code. Also added some functions readFields and readFlags which allows reading multiple
fields or flags in one by using loop and a single error check.

saxon-milton committed:
h264/bits/bitreader.go: added ByteAligned and BytesRead functions.
ByteAligned will return true if we are at the start of a byte. BytesRead returns the number of bytes we have read using this BitReader.

saxon-milton committed:
h264/nalUnit.go: removed usage of old BitReader and replaced with new

saxon-milton committed:
h264/pps.go: removed usage of old BitReader and replaced with new

saxon-milton committed:
h264/rbsp.go: removed usage of old BitReader and replaced with new

saxon-milton committed:
h264/server.go: removed usage of old BitReader and replaced with new

saxon-milton committed:
h264/slice.go: removed usage of old BitReader and replaced with new

saxon-milton committed:
h264/sps.go: removed usage of old BitReader and replaced with new

kortschak commented:
Don't use the word "resolves" in front of an issue unless you want the issue closed on merging the mentioning PR. GitHub doesn't do NLP, so the partly there is not understood. Use "updates" instead (at the bottom of the opening PR comment).

saxon-milton commented:
> Don't use the word "resolves" in front of an issue unless you want the issue closed on merging the mentioning PR. GitHub doesn't do NLP, so the partly there is not understood. Use "updates" instead (at the bottom of the opening PR comment).

Will do, cheers.

kortschak commented on h264/bits/bitreader.go:
```diff
@@ -155,3 +157,14 @@ func (br *BitReader) PeekBits(n int) (uint64, error) {
+
+// ByteAligned returns true if the reader position is at the start of a byte,
+// and false otherwise.
+func (br *BitReader) ByteAligned() bool {
```
There should be a test for this.

kortschak commented on h264/bits/bitreader.go:
```diff
@@ -155,3 +157,14 @@ func (br *BitReader) PeekBits(n int) (uint64, error) {
+}
+
+// BytesRead returns the number of bytes that have been read by the BitReader.
+func (br *BitReader) BytesRead() int {
```
This doesn't seem necessary. You can either query the underlying reader or just do the accounting by the caller.

saxon-milton replied:
So the caller should query the underlying reader for this information ? Is there an io.Reader implementation that provides this information ?

saxon-milton replied:
I think this method is stupid too, but I'm just trying not to break the current code. If I get rid of this, and we do the querying by the caller, there's going to be a lot more code added to this PR.

kortschak replied:
OK. Leave it as is. Why do we need to know the number of bytes that were read?

kortschak commented on h264/cabac.go:
```diff
@@ -1,6 +1,9 @@
-import "github.com/pkg/errors"
+import (
+	"github.com/ausocean/h264decode/h264/bits"
+	"github.com/pkg/errors"
```
See previous comments about introducing this as a dep. Note that mrmod's code does not import it.

kortschak reviewed (commented):
This doesn't look mostly mechanical. Is there a suggested review scheme?

saxon-milton commented:
> This doesn't look mostly mechanical. Is there a suggested review scheme?

Maybe I have a different (or incorrect) definition of mechanical. A lot of the changes in this PR are of this form,
`<some struct field> = b.NextField(<string name of field>,numberOfBits)`
to 
`b, err = br.ReadBits(numberOfBits)`
`if err != nil {`
`    return errors.Wrap(err,"could not read some field")`
`}`
`<some struct field> = int(b)`

Also, I've removed the old BitReader, and changed some of the old BitReaders methods to not be methods for a particular struct.

saxon-milton committed:
h264/bits/bitreader_test.go: added test for ByteAligned()
